### PR TITLE
Derive `Hash` implementations for tag unions

### LINKED ISSUE #4202: Implement derivers for Hash

### LINKED ISSUE #4206: Low-level to determine the discriminant of a tag union

### PR DESCRIPTION
**Pull request by ayazhafiz**
Implements auto-deriving of `Hash.hash` for recursive and non-recursive tag unions. Tag unions are hashed by first hashing their discriminant, and then hashing each value in the payload of the appropriate tag variant.

- Do not hash discriminants for tags that are newtypes. In such cases the discriminant doesn't exist, or can be seen as constant, so hashing it provides no value to the hash. (a308ebb38c8a206987a4f9a1b75ba90947b2b1a6)
- Implement hashing discriminants for tags that need them in the derived implementation itself, rather than relying on a LowLevel. There are two benefits to this - (1) we can choose a smaller byte size (u8) for the discriminant in the common case, rather than having to always hash the upper bound of discriminant sizes, and (2) it gets rid of a low-level that we otherwise don't need. (cb96a6425934d2e926a33e14e258efd030755644 and 792afe5457b71e7e748854a5719a1e5cd7872749)

Closes #4202
Closes #4206

### CODE REVIEW

File: crates/compiler/test_gen/src/gen_abilities.rs [hunk 1]
```diff
@@ -1372,5 +1372,228 @@ mod hash {
...
+                            |> Hash.hash c
+                            |> tRead
+                        "#
+                    ),
+                    TEST_HASHER,
+                ),
+                RocList::from_slice(&[0]),
+                RocList<u8>
+            )
+        }
```
Comment (ayazhafiz) [inline crates/compiler/test_gen/src/gen_abilities.rs:1597]: There are some unification problems here, will address in a future PR

File: crates/compiler/test_gen/src/gen_abilities.rs [hunk 2]
```diff
@@ -1372,5 +1372,228 @@ mod hash {
...
+                        "#
+                    ),
+                    TEST_HASHER,
+                ),
+                RocList::from_slice(&[
+                    0, 0, // dicsr A
+                    15, 23, // payloads A
+                    1, 0,  // discr B
+                    37, // payloads B
+                    2, 0, // discr B
```
Comment (ayazhafiz) [inline crates/compiler/test_gen/src/gen_abilities.rs:1563]:
```suggestion
                    2, 0, // discr C
```

File: crates/compiler/derive/src/hash.rs [hunk 3]
```diff
@@ -89,54 +103,326 @@ fn hash_record(env: &mut Env<'_>, fn_name: Symbol, fields: Vec<Lowercase>) -> (E
...
+    let union_sym = env.new_symbol("union");
+
+    let hasher_sym = env.new_symbol("hasher");
+    let hasher_var = synth_var(env.subs, Content::FlexAbleVar(None, Symbol::HASH_HASHER));
+
+    let (discr_width, discr_precision_var, hash_discr_member) = if union_tags.len() > u64::MAX as _
+    {
+        // Should never happen, `usize` isn't more than 64 bits on most machines, but who knows?
+        // Maybe someday soon someone will try to compile a huge Roc program on a 128-bit one.
+        internal_error!("new record unlocked: you fit more than 18 billion, billion tags in a Roc program, and the compiler didn't fall over! But it will now. 🤯")
```
Comment (rtfeldman) [inline crates/compiler/derive/src/hash.rs:171]: 😂

**Review by rtfeldman — APPROVED** [review]
Looks great! 😍😍😍

**Pull request merged**